Implement support for the ECSV format proposed in APE6

taldcroft opened the pull request:
APE6 proposes a new standard Data-table Text Interchange Format for storing data tables in a text-only format.  This PR provides a demonstration implementation of that in `astropy.io.ascii`.  This is by no means complete and should not be merged.

astrofrog commented:
@taldcroft - what is the status of this? Would it be fine to postpone until after 1.0 is release?

taldcroft commented:
@astrofrog - I was hoping to get it in to 1.0.  I've addressed all the review comments and nobody has expressed any objections to this or APE6.  I'm not sure if there are still high-level reservations in the concept or APE6 as it stands.

astrofrog commented:
@taldcroft - I haven't really had time to look over it properly due to lack of time with other issues. I will try and look today.

taldcroft commented:
Just rebased.  Curious, github was saying it couldn't merge but the rebase didn't have conflicts.

astrofrog commented:
It must have been a changelog entry. We now have a custom merge strategy for the changelog (see ``.gitattributes`` that makes it look like there is no conflict when rebasing).

taldcroft commented:
Ah, clever with the merge strategy!

cdeil commented:
:+1: to merge

astrofrog commented:
This looks good to me to and so does the APE. I do feel that the APE has not received enough review to be accepted/merged now as an APE, but I wonder if we can go ahead and merge this so that it gets some testing, and the finalize the APE later (and bump the ECSV version number if needed?)

@eteq @mdboom @embray - any thoughts on this?

embray commented:
I have no problem with including this provisionally, assuming the proper warnings are in place.

embray commented on CHANGES.rst:
```diff
@@ -91,6 +91,8 @@ New Features
   - Write comment lines in Table metadata during output for all basic formats,
     IPAC, and fast writers. This functionality can be disabled with
     ``comment=False``. [#3255]
```
An annoying nitpick--now there should be a blankline here (I was worried that the merge=union wouldn't always handle this properly, but it's still better than not having it at all.  Now if only it were possible to run custom merge strategies on GH....)

embray replied:
More substantively, while we're in here it might be worth nothing that the details of this format are provisional pending acceptance of APE<whatever>.  I definitely think we should advertise this more to get more community feedback, etc.

astrofrog commented:
@taldcroft - can the file format version be adjusted to somehow indicate that this is not the final version of the standard? ``0.9`` or something? :)

taldcroft commented:
It certainly could, and maybe that would make sense to have a 0.9 for a release in order to see any issues that come up in the wild.  But of course it would be nice to converge in the next couple of days so that hopefully no changes would be required in 1.0.  The issue I just raised would be a trivial change in the implementation.

Apart from that, after a relatively broad call for comment there haven't been any objections.

taldcroft commented:
Changed version to 0.9 and updated to use `datatype` instead of `columns` and `type`.

taldcroft commented:
Assuming tests pass, does anybody have additional comments, objections, or require further time for review for either APE-6 or this implementation?

mdboom commented:
I think there may be further changes to resolve the differences between ECSV and ASDF but (a) I think that's a bigger project than the 1.0 release schedule allows, and I don't think it's important enough to hold it up, and (b) it's probably possible to support the current syntax and the "resolved" ECSV/ASDF format indefinitely, so not extremely problematic.

astrofrog commented:
No objections to merging this as-is. I like @mdboom's suggestion of resolving the differences between ECSV and ASDF so maybe the APE itself should stay open for now?

taldcroft commented:
I think we need the APE merged at version 0.9 so there is an accepted document to reference.  Resolution of differences would go into the 1.0 version.

astrofrog commented:
Sounds good!

mhvk commented:
+1 to merging as 0.9, and seeing what comes up in actual usage. I'm looking forward to starting to using this myself!

astrofrog commented:
@taldcroft - there are failures related to the fact that the tests should pass or be skipped if pyyaml is not installed. In addition, pyyaml should be installed for the optional dependency build(s) in ``.travis.yml`` and should be installed by default in ``appveyor.yml``. Note that it can be installed with ``conda``.

eteq commented:
APE6 has been accepted, so I'm merging this.  Thanks @taldcroft !

astrofrog commented:
Thanks @taldcroft! :tada:

taldcroft commented:
Long live the meta!
